### PR TITLE
feat: add NodeIdNodeCollection for identifying nodes by "node IDs"

### PR DESCRIPTION
**Pull request by larsmoa**
# Description

Added NodeIdNodeCollection for identifying nodes by "node IDs". This is similar to TreeIndexNodeCollection, but can be useful for cases where node IDs are known, but tree index and subtree size is not.

Also renamed `registerCustomNodeCollectionType` to `registerNodeCollectionType` as this should be the only mechanism for registering `NodeCollection` (and not have the god-like object `NodeCollectionDeserializer` that knows about them all).

# Checklist:

Here is a checklist that should completed before merging this given feature. 
Any shortcomings from the items below should be explained and detailed within the contents of this PR.

- [x] I am proud of this feature.
- [x] I have performed a self-review of my own code.
- [x] I have manually tested this for different scenarios (different models, formats, devices, browsers).
- [x] I have commented my code in hard-to-understand areas.
- [x] I have made corresponding changes to the public documentation.
- [x] I have added unit and visuals tests to prove that my feature works to the best of my ability.
- [x] I have refactored the code for readability to the best of my ability.
- [x] I have checked that my changes do not introduce regressions in the public documentation.
- [x] I have outlined any known defects / lacking capabilities in the contents of this PR.
- [x] I have listed any remaining work that should follow this PR in the description or jira/miro/etc.
- [x] I have added TSDoc to any public facing changes.
- [x] I have added "developer documentation" in any package README.md that is related / applicable to this PR.
- [x] I have noted down and am currently tracking any technical debt introduced in this PR.
- [x] I have thoroughly thought about the architecture of this implementation.
- [ ] I have asked peers to test this feature.

### CODE REVIEW

File: viewer/packages/cad-styling/src/NodeIdNodeCollection.ts [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+import { NodeCollectionSerializationContext, registerNodeCollectionType } from './NodeCollectionDeserializer';
+import { SerializedNodeCollection } from './SerializedNodeCollection';
+
+/**
+ * Collection that holds a set of nodes including children identified by nodeIds. Note that
+ * this involves remapping IDs to "tree indices" and subtree sized used by Reveal using
+ * the CDF API. Its often better to use one of the other node collections or {@see TreeIndexNodeCollection}
+ * whenever possible for best performance.
+ */
+export class NodeIdNodeCollection extends CdfNodeCollectionBase {
```
Comment (pramodcog) [inline viewer/packages/cad-styling/src/NodeIdNodeCollection.ts:17]: `NodeCollectionOfNodeId` could be better name?

File: viewer/packages/cad-styling/src/NodeIdNodeCollection.ts [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+  private _nodeIds: number[] = [];
+
+  constructor(client: CogniteClient, model: CdfModelNodeCollectionDataProvider) {
+    super(NodeIdNodeCollection.classToken, model);
+    this._client = client;
+    this._model = model;
+  }
+
+  /**
+   * Populates the collection with the nodes with the IDs provided. All children of
```
Comment (pramodcog) [inline viewer/packages/cad-styling/src/NodeIdNodeCollection.ts:32]:
```suggestion
   * Populates the collection with the nodes from the IDs provided. All children of
```

File: viewer/packages/cad-styling/src/NodeIdNodeCollection.ts [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+    const executeRequest = async (batchNodeIds: number[]) => {
+      const response = await this._client.revisions3D.retrieve3DNodes(
+        modelId,
+        revisionId,
+        batchNodeIds.map(id => ({ id }))
+      );
+      const listResponse = new SingleRequestListResponse<Node3D[]>(response);
+      return listResponse;
+    };
+    const responses = chunk(nodeIds, 1000).map(chunk => executeRequest(chunk));
```
Comment (pramodcog) [inline viewer/packages/cad-styling/src/NodeIdNodeCollection.ts:54]: Why smaller array, any significance with 1000 number here?

Comment (larsmoa) [reply viewer/packages/cad-styling/src/NodeIdNodeCollection.ts:54]: Yes, this is a limitation in the API - see https://docs.cognite.com/api/v1/#tag/3D-Model-Revisions/operation/get3DNodesById

File: viewer/packages/cad-styling/src/NodeIdNodeCollection.ts (hunk 1 above)
Comment (larsmoa) [reply viewer/packages/cad-styling/src/NodeIdNodeCollection.ts:17]: I tried to keep the style we have today - e.g. `TreeIndexNodeCollection`. Could probably consider all collections really...

File: viewer/packages/cad-styling/src/NodeIdNodeCollection.ts (hunk 2 above)
Comment (larsmoa) [reply viewer/packages/cad-styling/src/NodeIdNodeCollection.ts:32]: Hmm, to me it makes more sense with 'with'. @christjt

File: viewer/packages/cad-styling/src/NodeIdNodeCollection.test.ts [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+    mockRevisions3DAPI.verify(x => x.retrieve3DNodes(It.IsAny(), It.IsAny(), It.IsAny()), Times.Exactly(2));
+  });
+
+  test('executeFilter with 0 nodes, does not trigger any requests', async () => {
+    const collection = new NodeIdNodeCollection(mockClient.object(), mockModel.object());
+    await collection.executeFilter([]);
+    mockRevisions3DAPI.verify(x => x.retrieve3DNodes(It.IsAny(), It.IsAny(), It.IsAny()), Times.Never());
+  });
+
+  test('executeFilter maps node bounds to ThreeJS coordinates', async () => {
```
Comment (pramodcog) [inline viewer/packages/cad-styling/src/NodeIdNodeCollection.test.ts:48]: Is the test verifying  `executeFilter maps node bounds to ThreeJS coordinates`? I think it is validating collection `IndexSet` & `Areas` values, does not sounds to be related test description :grey_question:

File: viewer/packages/cad-styling/src/SinglePropertyFilterNodeCollection.ts [hunk 5]
```diff
@@ -72,35 +59,20 @@ export class SinglePropertyFilterNodeCollection extends NodeCollection {
    * @param propertyKey Node property key, e.g. `':FU'`.
    * @param propertyValues Lookup values, e.g. `["AR100APG539","AP500INF534","AP400INF553", ...]`
    */
-  async executeFilter(propertyCategory: string, propertyKey: string, propertyValues: string[]): Promise<void> {
+  executeFilter(propertyCategory: string, propertyKey: string, propertyValues: string[]): Promise<void> {
```
Comment (pramodcog) [inline viewer/packages/cad-styling/src/SinglePropertyFilterNodeCollection.ts:62]: Isn't this function need `async`?

Comment (pramodcog) [reply viewer/packages/cad-styling/src/SinglePropertyFilterNodeCollection.ts:62]: Good to know

**Review by pramodcog — APPROVED** [review]

**Pull request merged**